src/python/pants/backend/adhoc/code_quality_tool.py

```diff
@@ -0,0 +1,549 @@
+from pants.engine.rules import Rule, collect_rules, rule
+from pants.engine.target import (
+    COMMON_TARGET_FIELDS,
+    FieldSetsPerTarget,
+    FieldSetsPerTargetRequest,
+    SpecialCasedDependencies,
+    StringField,
+    StringSequenceField,
+    Target,
+    Targets,
+)
+from pants.option.option_types import SkipOption
+from pants.option.subsystem import Subsystem
+from pants.util.frozendict import FrozenDict
+from pants.util.strutil import Simplifier, help_text
+
+
+class CodeQualityToolFileGlobIncludeField(StringSequenceField):
+    alias: ClassVar[str] = "file_glob_include"
+    required = True
+    help = help_text(
+        """
+        Globs that identify files that can be processed by this tool
+
+        A file matching any of the supplied globs is eligible for processing.
+        Example: ["**/*.py"]
+        """
+    )
+
+
+class CodeQualityToolFileGlobExcludeField(StringSequenceField):
+    alias: ClassVar[str] = "file_glob_exclude"
+    required = False
+    default = ()
+    help = help_text(
+        """
+        Globs matching files that should not be processed by this tool
+        """
+    )
+
+
+class CodeQualityToolRunnableField(StringField):
+    alias: ClassVar[str] = "runnable"
+    required = True
+    help = help_text(
+        lambda: f"""
+        Address to a target that can be invoked by the `run` goal (and does not set
+        `run_in_sandbox_behavior=NOT_SUPPORTED`). This will be executed along with any arguments
+        specified by `{CodeQualityToolArgumentsField.alias}`, in a sandbox with that target's transitive
+        dependencies, along with the transitive dependencies specified by
+        `{CodeQualityToolExecutionDependenciesField.alias}`.
+        """
+    )
+
+
+class CodeQualityToolArgumentsField(StringSequenceField):
+    alias: ClassVar[str] = "args"
+    default = ()
+    help = help_text(
+        lambda: f"""
+        Extra arguments to pass into the `{CodeQualityToolRunnableField.alias}` field
+        before the list of source files
+        """
+    )
+
+
+class CodeQualityToolExecutionDependenciesField(SpecialCasedDependencies):
+    alias: ClassVar[str] = "execution_dependencies"
+    required = False
+    default = None
+
+    help = help_text(
+        """
+        Additional dependencies that need to be available when running the tool.
+        Typically used to point to config files.
+        """
+    )
+
+
+class CodeQualityToolRunnableDependenciesField(SpecialCasedDependencies):
+    alias: ClassVar[str] = "runnable_dependencies"
+    required = False
+    default = None
+
+    help = help_text(
+        lambda: f"""
+        The runnable dependencies for this command.
+
+        Dependencies specified here are those required to exist on the `PATH` to make the command
+        complete successfully (interpreters specified in a `#!` command, etc). Note that these
+        dependencies will be made available on the `PATH` with the name of the target.
```

Comment:
This is an interesting aspect of the existing adhoc_tool machinery, which is not without its gotchas (I ran into it when setting up the demo). Its not introduced in this PR and the documentation is taken straight from `AdhocToolRunnableDependenciesField`. Here's how it goes FYI:

consider this setup for `markdownlint` which requires `node` to run:
```python
system_binary(
    name="node",    # This is the name that the binary will be available as in the sandbox
    binary_name="node",    # not this!
    # fingerprint stuff omitted
)

system_binary(
    name="markdownlint",
    binary_name="markdownlint",
    # fingerprint stuff omitted
)

code_quality_tool(
    name="markdownlint_linter",
    runnable=':markdownlint',   #
    runnable_dependencies=[":node"]   # Needs to be available on the PATH when mardkownlint binary runs.
)
```
Put in other words, if we had set:
```python
system_binary(
    name="node_bin",    # Not plain `node`
    binary_name="node", 
    # fingerprint stuff omitted
)

code_quality_tool(
    name="markdownlint_linter",
    runnable=':markdownlint',   #
    runnable_dependencies=[":node_bin"] 
)
```
the code quality tool won't work because, even though node is put in the sandbox, it is available to run by the name `node_bin` only and `markdownlint` won't be able to find it.

Reply:
that makes sense, thanks for the explanation. I can't really think of a clearer way of phrasing that, and at least it's consistent with adhoc